File: Source/JS/src/Client/MDSDocument.js
//
//  MDSDocument.js
//
//  Created by Stevo on 5/18/2022.
//  Copyright © 2022 Stevo Brock. All rights reserved.
//

// Imports
let	uuid = require('uuid');
let	{UuidTool} = require('uuid-tool');

//----------------------------------------------------------------------------------------------------------------------
// MDSDocument
module.exports = class MDSDocument {

	// Properties
	get documentID() { return this.info.documentID; }							// String
	get	revision() { return this.info.revision; }								// Integer (null if not yet created)

	get isActive() { return this.info.active == 1; }							// Bool
	setActive(active) {
			// Update
			this.info.active = active ? 1 : 0;
			
			// Note updated
			this.activeChanged = true;
		}
	toggleIsActive() {
			// Update
			this.info.active = 1 - this.info.active;
		
			// Note updated
			this.activeChanged = true;
		}

	get	creationDate() { return new Date(this.info.creationDate); }				// Date (null if not yet created)
	get	modificationDate() { return new Date(this.info.modificationDate); }		// Date (null if not yet created)

	// Lifecycle methods
	//------------------------------------------------------------------------------------------------------------------
	constructor(info) {
		// Check situation
		if (info == null) {
			// New
			let uuidBytes = UuidTool.toBytes(UuidTool.newUuid());
			let documentID = btoa(String.fromCharCode(...new Uint8Array(uuidBytes))).slice(0, 22);

			this.info = {documentID: documentID, active: 1, json: {}, attachments: {}};
			this.isNew = true;
		} else if ((typeof info) == 'string') {
			// Decode
			this.info = JSON.parse(info);
			this.isNew = false;
		} else {
			// Store
			this.info = info;
			this.isNew = false;
		}
		
		// Setup
		this.updated = {};
		this.removed = new Set();
		this.activeChanged = false;
	}

	// Instance methods
	//------------------------------------------------------------------------------------------------------------------
	attachments(type) {
		// Return attachments filtered by given type
		return Object.fromEntries(
				Object.entries(this.info.attachments)
						.filter(([id, info]) => info.info.type == type));
	}

	//------------------------------------------------------------------------------------------------------------------
	toString() { return JSON.stringify(this.info); }

	// Subclass methods
	//------------------------------------------------------------------------------------------------------------------
	value(property) {
		// Check situation
		if (property in this.updated)
			// Have updated value
			return this.updated[property];
		else if (this.removed.has(property))
			// Was removed
			return null;
		else
			// Not updated
			return this.info.json[property];
	}

	//------------------------------------------------------------------------------------------------------------------
	set(property, value = null) {
		// Check value
		if (value != null) {
			// Have value
			this.updated[property] = value;
			this.removed.delete(property);
		} else {
			// Don't have value
			delete this.updated[property];
			this.removed.add(property);
		}
	}

	//------------------------------------------------------------------------------------------------------------------
	notePropertyChanged(property) {
		// Updated
		this.updated[property] = this.info.json[property];
		this.removed.delete(property);
	}

	// Internal methods
	//------------------------------------------------------------------------------------------------------------------
	hasCreateInfo() { return !self.creationDate; }

	//------------------------------------------------------------------------------------------------------------------
	createInfo() {
		// Return info
		return {
			documentID: this.info.documentID,
			json: this.updated,
		};
	}

	//------------------------------------------------------------------------------------------------------------------
	updateFromCreate(info) {
		// Update
		this.info.revision = info.revision;
		this.info.creationDate = info.creationDate;
		this.info.modificationDate = info.modificationDate;
		this.info.json = this.updated;

		this.isNew = false;

		// Reset
		this.updated = {};
		this.removed.clear();
	}

	//------------------------------------------------------------------------------------------------------------------
	hasUpdateInfo() { return (Object.keys(this.updated).length > 0) || (this.removed.size > 0) || this.activeChanged; }

	//------------------------------------------------------------------------------------------------------------------
	updateInfo() {
		// Setup
		let	updateInfo = {documentID: this.info.documentID};

		// Check if have updated properties
		if (Object.keys(this.updated).length > 0)
			// Have updated properties
			updateInfo.updated = this.updated;
		
		// Check if have removed properties
		if (this.removed.size > 0)
			// Have removed properties
			updateInfo.removed = Array.from(this.removed);
		
		// Check if active changed
		if (this.activeChanged)
			// Active changed
			updateInfo.active = this.info.active;
		
		return updateInfo;
	}

	//------------------------------------------------------------------------------------------------------------------
	updateFromUpdate(info) {
		// Update
		this.info.revision = info.revision;
		this.info.active = info.active;
		this.info.modificationDate = info.modificationDate;

		this.info.json = info.json;

		// Reset
		this.updated = {};
		this.removed.clear();
		this.activeChanged = false;
	}
}

File: Source/Python/src/mini_document_storage/client/MDSDocument.py
#
# MDSDocument.py
#
# Created by Stevo on 5/16/2023.
# Copyright © 2023 Stevo Brock.  All rights reserved
#

# Imports
import base64
from datetime import datetime
import uuid

#-----------------------------------------------------------------------------------------------------------------------
# MDSDocument
class MDSDocument:
    
	# Properties
	@property
	def document_id(self):																# str
		return self.info['documentID']
	@property
	def revision(self):																	# int (None if not yet created)
		return self.info.get('revision')
	
	@property
	def is_active(self):																# bool
		return self.info['active'] == 1
	def set_active(self, active):
		# Update
		self.info['active'] = 1 if active else 0
	
		# Note updated
		self.active_changed = True
	def toggle_is_active(self):
		# Update
		self.info['active'] = 1 - self.info['active']

		# Note updated
		self.active_changed = True
	
	@property
	def creation_date(self):															# datetime (None if not yet created)
		value = self.info.get('creationDate')

		return datetime.fromisoformat(value.replace('Z', '+00:00')) if value else None
	
	@property
	def modification_date(self):														# datetime (None if not yet created)
		value = self.info.get('modificationDate')

		return datetime.fromisoformat(value.replace('Z', '+00:00')) if value else None
	
	@property
	def	has_create_info(self):															# bool
		return self.creation_date == None
	
	@property
	def has_update_info(self):															# bool
		return self.updated or self.removed or self.active_changed
	
	# Lifecycle methods
	#-------------------------------------------------------------------------------------------------------------------
	def __init__(self, info = None):
		# Check situation
		if info == None:
			# New
			document_id = base64.urlsafe_b64encode(uuid.uuid4().bytes).decode('utf-8').strip('=')

			self.info = {'documentID': document_id, 'active': 1, 'json': {}, 'attachments': {}}
			self.is_new = True
		else:
			# Store
			self.info = info
			self.is_new = False
		
		# Finish setup
		self.updated = {}
		self.removed = set()
		self.active_changed = False
	
	# Python methods
	#-------------------------------------------------------------------------------------------------------------------
	def __eq__(self, other):
		return self.document_id == other.document_id

	#-------------------------------------------------------------------------------------------------------------------
	def __hash__(self):
		return hash(self.document_id)

	# Instance methods
	#-------------------------------------------------------------------------------------------------------------------
	def attachments(self, type):
		return {id: info for (id, info) in self.info['attachments'].items() if info['info']['type'] == type}
	
	# Subclass methods
	#-------------------------------------------------------------------------------------------------------------------
	def value(self, property):
		# Check situation
		if property in self.updated:
			# Have updated value
			return self.updated[property]
		elif property in self.removed:
			# Was removed
			return None
		elif property in self.info['json']:
			# Not updated
			return self.info['json'][property]
		else:
			# Not present
			return None

	#-------------------------------------------------------------------------------------------------------------------
	def set(self, property, value = None):
		# Check value
		if value != None:
			# Have value
			self.updated[property] = value
			self.removed.discard(property)
		else:
			# Don't have value
			self.updated.pop(property, None)
			self.removed.add(property)

	#-------------------------------------------------------------------------------------------------------------------
	def note_property_changed(self, property):
		# Updated
		self.updated[property] = self.info['json'][property]
		self.removed.discard(property)

	# Internal methods
	#-------------------------------------------------------------------------------------------------------------------
	def create_info(self):
		# Return info
		return {'documentID': self.document_id, 'json': self.updated}

	#-------------------------------------------------------------------------------------------------------------------
	def update_from_create(self, info):
		# Update
		self.info['revision'] = info['revision']
		self.info['creationDate'] = info['creationDate']
		self.info['modificationDate'] = info['modificationDate']
		self.info['json'] = self.updated

		self.is_new = False

		# Reset
		self.updated = {}
		self.removed.clear()

	#-------------------------------------------------------------------------------------------------------------------
	def update_info(self):
		# Setup
		update_info = {'documentID': self.document_id}

		# Check if have updated properties
		if self.updated:
			# Have updated properties
			update_info['updated'] = self.updated
		
		# Check if have removed properties
		if self.removed:
			# Have removed properties
			update_info['removed'] = list(self.removed)
		
		# Check if active changed
		if self.active_changed:
			# Active changed
			update_info['active'] = self.info['active']
		
		return update_info

	#-------------------------------------------------------------------------------------------------------------------
	def update_from_update(self, info):
		# Update
		self.info['revision'] = info['revision']
		self.info['active'] = info['active']
		self.info['modificationDate'] = info['modificationDate']

		self.info['json'] = info['json']

		# Reset
		self.updated.clear()
		self.removed.clear()
		self.active_changed = False
